keep stacktrace in abstract_conv optimizations

vmichals opened the pull request:
fix #3589

lamblin commented on theano/tensor/nnet/opt.py:
```diff
@@ -92,6 +94,7 @@ def local_abstractconv_gradweight_gemm(node):
     if node.op.filter_flip:
```
I think we should also add the stack trace on the output of `CorrMM_gradWeights`, the `rval` defined on l. 91.

lamblin replied:
(done)

lamblin commented on theano/tensor/nnet/opt.py:
```diff
@@ -175,6 +181,7 @@ def local_conv2d_gradweight_cpu(node):
         rval = rval[:, :, ::-1, ::-1]
         rval = theano.tensor.patternbroadcast(rval,
                                               node.outputs[0].broadcastable)
+        copy_stack_trace(node.outputs[0], rval)
```
I would keep this one, but also add the stack trace to the output of `convGrad3D`, defined on l.172.

lamblin replied:
(done)

lamblin commented on theano/tensor/nnet/opt.py:
```diff
@@ -245,6 +252,8 @@ def local_conv2d_gradweight_cpu(node):
 
     res = theano.tensor.patternbroadcast(res, node.outputs[0].broadcastable)
+
+    copy_stack_trace(node.outputs[0], res)
```
Again, this one is fine, but I would also add the trace to the `res` defined on line 249.

lamblin replied:
(done)

lamblin commented on theano/tensor/nnet/opt.py:
```diff
@@ -278,6 +287,8 @@ def local_conv2d_gradinputs_cpu(node):
         rval = theano.tensor.patternbroadcast(rval,
                                               node.outputs[0].broadcastable)
+
+        copy_stack_trace(node.outputs[0], rval)
```
That one is fine as well, just add it to the output of `convTransp3D` as well.

lamblin replied:
(done)

lamblin commented on theano/tensor/nnet/opt.py:
```diff
@@ -327,6 +338,7 @@ def local_conv2d_gradinputs_cpu(node):
                  direction_hint='bprop inputs')
     din = din(topgrad, filters)
```
Let's add the stack trace here as well.

lamblin replied:
(done)

lamblin commented on theano/sandbox/gpuarray/tests/test_abstractconv.py:
```diff
@@ -75,6 +75,7 @@ def run_fwd(self, inputs_shape, filters_shape, ref=dnn_conv,
                         filter_shape=kshp)
+        self.assertTrue(hasattr(c.tag, 'trace'))
         f_ref = theano.function([], c_ref, mode=mode)
         f = theano.function([], c, mode)
```
Actually, you should check the presence of the stack trace on the output of the optimized graph, not the original variable.
The variable corresponding to the optimized output would be `f.maker.fgraph.outputs[0]`, and we want `f.maker.fgraph.outputs[0].tag.trace` to be there.

lamblin commented on theano/sandbox/gpuarray/tests/test_abstractconv.py:
```diff
@@ -113,6 +114,7 @@ def run_gradweight(self, inputs_shape, filters_shape, output_shape,
+        self.assertTrue(hasattr(c.tag, 'trace'))
         c = c(inputs, output, filters_shape[-2:])
         c_ref = ref(inputs, output,
                     filters_shape,
```
Same here (after defining `f` on line 124).

lamblin commented on theano/sandbox/gpuarray/tests/test_abstractconv.py:
```diff
@@ -159,6 +161,7 @@ def run_gradinput(self, inputs_shape, filters_shape, output_shape, ref=dnn_gradi
                                            subsample=subsample,
                                            filter_flip=filter_flip,
                                            imshp=imshp, kshp=kshp)
+        self.assertTrue(hasattr(c.tag, 'trace'))
```
Same here (`f` on line 169).

lamblin commented:
Hmm, you seemed to have had some trouble rebasing. We can try to solve that tomorrow if you want.

vmichals commented:
I think I fixed it by hard resetting to the previous commit. Then I rebased and resolved a minor merge conflict (keeping changes from another commit, that was made after branching).

lamblin commented:
The rebase and PR look good, thanks.
I'm ready to merge when Travis passes.
